### DISCUSSION TITLE
Websocket device

### DISCUSSION THREAD

**Discussion by glassTransition** [artisan-roaster-scope/artisan#701] [Q&A] [answered]
I'm attempting to implement a DIY thermocouple device (that I will document and share) with a raspberry pi zero W and would appreciate it if someone can share more details about the Websockets device interface. Specifically, captured messages would be very useful. The documentation outlines the general form but it would help to see actual traffic.

**Comment by glassTransition** [comment]
I just installed the latest version (2.4.6) and Websockets are gone as a device type?

**Comment by MAKOMO** [comment]
WebSocket support has not been touched in v2.4.6 and is still there. Documentation is here: <https://artisan-scope.org/devices/websockets/> and the details of messages can be configured under menu Config >> Ports, last tab "Web Sockets".

**Comment by glassTransition** [comment]
I was looking for the Websocket setup under the device rather than the port. Thank you for the clarification.

I have configured like the Documentation shows with the beanTemp and ambientTemp on inputs 1 and 2. How are those bound to BT and ET? 

If I click the "on" button in Artisan, it connects to my server, but I don't see any requests for temperatures. Are those supposed to be pushed to Artisan?

**Reply by MAKOMO** [reply]
Why not attach your configuration, best as Artisan settings .aset file (renamed to extension .txt)? Hard to help at this level. The WebSocket support in Artisan follows a request-response pattern and was introduced to support the Probat Series III machines. Maybe it is worse to look at the corresponding setup.

**Comment by glassTransition** [comment]
Thank you for your help with this.

Based on your comments I selected the Probat sample roaster as the machine type. I am actually planning to add my websocket temperature sensors to a Behmor 1600 which is of course not a choice in the machines list. 

The aset file in the machines folder did not update though.  Is that the file useful to see then? Is there another file with my settings?

I am temporarily replacing my server with a node-red flow so I can see all details of the traffic.

**Reply by MAKOMO** [reply]
> Thank you for your help with this.
> 
> Based on your comments I selected the Probat sample roaster as the machine type. I am actually planning to add my websocket temperature sensors to a Behmor 1600 which is of course not a choice in the machines list.
> 
> The aset file in the machines folder did not update though. Is that the file useful to see then? Is there another file with my settings?
> 
> I am temporarily replacing my server with a node-red flow so I can see all details of the traffic.

You can save your setting into an artisan-settings.aset file via menu Help >> Save Settings. To attach it here you need to rename the file such that it has a .txt extension.

**Comment by glassTransition** [comment]
Thank you. 

Here is my settings file:
[artisan-settings.txt](https://github.com/artisan-roaster-scope/artisan/files/7238524/artisan-settings.txt)

**Reply by MAKOMO** [reply]
Looks good to me. Does it work?

**Comment by glassTransition** [comment]
I'm making progress. This is a new area for me so I am learning as I debug. So far I'm getting:

[WS] Connection was opened.
[WS] Incoming message: {"command": "getCurrentRoastingStep", "id": 40922, "roasterID": 0}
[WS] Incoming message: {"command": "keepAlive", "id": 8371, "roasterID": 0}
[WS] Incoming message: {"command": "getCurrentRoastingStep", "id": 35474, "roasterID": 0}
[WS] Incoming message: {"command": "getCurrentRoastingStep", "id": 96767, "roasterID": 0}

Can I create a "custom machine" that only reports BT and ET so that I don't get these queries for functions that I don't have?

**Reply by MAKOMO** [reply]
Well, I don't know which functions you have. Maybe its time for you too check out the source code of Artisan for details. The kepAlive query is set in the Probat setup that you took as starting point. Just remove that in the Events tab (under Sampling) if you do not need that. Same for "getCurrentRoastingStep", which is set in the WebSocket tab under Commands.

**Comment by glassTransition** [comment]
Ok, thank you. I see the events that I don't need and can disable them. I don't see the ones that I do want though.

All I need are events to get the environmental temperature and bean temperature so that Artisan can plot them and calculate RoR and DT/ dt etc.

I could make up my own events, but I still don't know what to commands to use and how / where the data that my server returns is bound to the "special" variables ET and BT?

**Comment by glassTransition** [comment]
Looking at the source code, I would guess that temp1 is ET and temp2 is BT by default. So are those the variable names I should return with the temperature values in the JSON response string?

What command will Artisan send?

**Comment by glassTransition** [comment]
On the device assignment dialog when I select the ET / BT tab and the "Meter" radio button and "WebSocket" in the dialog, I have the "ET" and "BT" boxes checked for Curves. So are these the variable names in the JSON name-value pairs that I should return? What command should be sent by Artisan to request them? Do I have to create an event or message somehow through the Ports Configuration Websocket dialog?

**Comment by glassTransition** [comment]
I set the events as shown, however, I'm still getting the getCurrentRoastingStep command. I don't think I understand something.

![image](https://user-images.githubusercontent.com/4940369/135100403-8c177609-1102-49d3-946a-3a88841bdb88.png)

**Comment by MAKOMO** [comment]
This goes beyond what I can support for free. Please book a support hour with me if you need my further advice on this: https://shop.artisan.plus/en/consulting/video-conference

**Comment by glassTransition** [comment]
This should be a basic question, but if you can't help I understand. Thank you for your help so far.

**Comment by glassTransition** [comment]
I have no commercial interest btw. I am just trying to expand your user base by making it easier to instrument hobbyist roasters.

**Comment by CarefreeBuzzBuzz** [comment]
Considering how many questions, doesn’t seem basic.  You are freee to book consulting time.  

Michael Herbert 
CarefreeBuzzBuzz

**Reply by glassTransition** [reply]
It's really the same two questions that I've tried to ask multiple ways since I haven't gotten an answer. 

1) What do I need to call the variables in the message response so that Artisan knows they are ET / BT?

2) What command asks for them?

**Reply by poundy** [reply]
It's really already documented for you though. I know *nothing* about websockets, but from the previously linked documentation page https://artisan-scope.org/devices/websockets/, you can see all the things you need.
1. You call them whatever you fill in via the NODE value on the dialog box shown in the page above.  
2. The request comes from the Commands, Data Request value. Put whatever value there you need. 

If you read the detail, GetCurrentRoastingStep is the command that is issued to retrieve data - you can change it as per point 2 above, so that it works for whatever you have called your function.  If you find ET and BT are transposed, you can change them as per point 1. 

hope this helps you cross the current valley to reach the next !

**Reply by poundy** [reply]
also there's detail in there saying how Artisan handles if you need to call a specific webrequest to get just the BT or just the ET (or just any value) by having the "request" field on an input object as non-blank. Perhaps that's what you're expecting, but as documented, it's more efficient to use a single request to retrieve all values whenever possible

**Comment by MAKOMO** [comment, marked as answer]
I extended the documentation by a simple example on how to harvest ET and BT via WebSockets: https://artisan-scope.org/devices/websockets/

**Reply by glassTransition** [reply]
Thank you Marko. That's what I needed.

My server works now. I have attached the python script I'm using for the server running on a raspberry pi zero w. I will add reading of actual thermocouples and update it in the future.

[artisan_websocket_server.txt](https://github.com/artisan-roaster-scope/artisan/files/7251100/artisan_websocket_server.txt)
